Optimize the SIMD StringSearcher by searching first two chars

jiebinn opened the pull request:
This patch offers the optimized SIMD StringSearcher by searching the first and second chars together rather than only the first char, which will result in big performance gain. The patch also provides a quick path when the needle size is 1.

With this patch, I have tested the 43 queries in clickbench on ICX server. Query 20 has got 35% performance gain. Other StringSearcher related queries have got around 10% performance improvement. And the overall geomean of all the queries has got 4.1% performance gain.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Optimize the generic SIMD StringSearcher by searching first two chars.

jiebinn committed:
Optimize the SIMD StringSearcher by searching first two chars

This patch offers the optimized SIMD StringSearcher by searching the first
and second chars together rather than only the first char, which will result
in big performance gain. The patch also provides a quick path when the needle
size is 1.

With this patch, I have tested the 43 queries in clickbench on ICX server.
Query 20 has got 35% performance gain. Other StringSearcher related queries
have got around 10% performance improvement. And the overall geomean of all
the queries has got 4.1% performance gain.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn commented:
I will leave the avx512 optimization for another PR with more performance data.

alexey-milovidov commented:
The results are very promising!
Need to also check if there are no degradations on long needles, as in that case, Volnitsky can be theoretically better.

jiebinn commented:
> The results are very promising! Need to also check if there are no degradations on long needles, as in that case, Volnitsky can be theoretically better.

Hi Alexey, Thanks for your advice. I have read the Volnitsky algorithm, and it would have a good performance if the needle_size is big. Then I designed the tests for difference needle_size. 

From the result, the optimized SSE algorithm would have a better performance if the needle_size <= 20. For example, there is 42% gain searching for ’oogle‘ or '35%' gain searching for 'google'. 

But there could be at most 9%-10% regression than Volnitsky algorithm when the needle_size is large. We might set a threshold between the SSE and Volnitsky currently. We could remove the threshold if the optimized SIMD algorithm wins all the time.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jiebinsu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{border:.5pt solid windowtext;}
.xl66
	{text-align:center;
	border:.5pt solid windowtext;}
.xl67
	{mso-number-format:"0\.0%";
	border:.5pt solid windowtext;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



Query:   SELECT COUNT(*) FROM hits WHERE URL LIKE '%Needle%';
--
Needle | Needle_size | opt/baseline
oogle | 5 | 142.0%
m=google | 8 | 118.4%
com=google | 10 | 109.1%
?com=google.ru/arts/ | 20 | 101.9%
p?com=google.ru/arts/ | 21 | 99.6%
google.ru/arts/searchAuto | 25 | 97.8%
p://alpari.ru/images/0000   до 110.html_params | 44 | 90.6%
ges/0000 до   110.html_params%3Drhost%3D48%2F&ti=Покупки вампирог&utf8=✓ | 70 | 90.9%
%2F%2Fwwwww.bonprix.ru&pvid=16853217/Tajmery_kangoo_balances/news.googlead%26ar_passenger | 89 | 92.8%
alpari.ru/images/0000   до 110.html_params%3Drhost%3D48%2F&ti=Покупки   вампирог&utf8=✓&search?text=фильм тихоокеанский-google.ru | 125 | 90.1%
images/0000 до   110.html_params%3Drhost%3D48%2F&ti=Покупки   вампирог&utf8=✓&search?text=фильм   тихоокеанский-google.ru/catalog=on&acce.4077294 | 139 | 89.9%



</body>

</html>

jiebinn commented:
> The results are very promising! Need to also check if there are no degradations on long needles, as in that case, Volnitsky can be theoretically better.

Hi @alexey-milovidov , Thanks for your advice. I have read the Volnitsky algorithm, and it would have a good performance if the needle_size is big. Then I designed the tests for difference needle_size.

From the result, the optimized SSE algorithm would have a better performance if the needle_size <= 20. For example, there is 42% gain searching for ’oogle‘ or '35%' gain searching for 'google'.

But there could be at most 7.8% regression than Volnitsky algorithm when the needle_size is large, such as 156. We might set a threshold between the SSE and Volnitsky currently. 

![image](https://user-images.githubusercontent.com/49742809/220025743-957e1049-aa88-485a-82fe-d22309478871.png)

jiebinn commented:
In the chart, the large needle has not shown due to limited space. 

Needle_size = 94
Needle = 'http:%2F%2Fwwwww.bonprix.ru&pvid=16853217/Tajmery_kangoo_balances/news.googlead%26ar_passenger'

Needle_size = 156
Needle = 'http://alpari.ru/images/0000 до 110.html_params%3Drhost%3D48%2F&ti=Покупки вампирог&utf8=✓&search?text=фильм тихоокеанский-google.ru/catalog=on&acce.4077294'

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -587,9 +589,14 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+        first_pattern = _mm_set1_epi8(first);
+        if (needle + 1 < needle_end)
+        {
+            second  = *(needle + 1);
```
(cosmetic: superfluous space after second)

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -569,8 +569,10 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
 #ifdef __SSE4_1__
-    /// vector filled `first` for determining leftmost position of the first symbol
-    __m128i pattern;
+    uint8_t second{};
```
Let's add a comment before l. 573.: `/// second character of "needle" (if it's length is > 1)`.

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -569,8 +569,10 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
 #ifdef __SSE4_1__
-    /// vector filled `first` for determining leftmost position of the first symbol
-    __m128i pattern;
+    uint8_t second{};
```
`first` and `second` are too generic as names. Should have been `first_needle_character` and `second_needle_character`.

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -569,8 +569,10 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
-    /// vector filled `first` for determining leftmost position of the first symbol
-    __m128i pattern;
+    uint8_t second{};
+    /// vector filled `first` or `second` for determining leftmost position of the first and second symbols
```
Variable names `first_pattern` and `second_pattern` are too generic. Let's change them to `first_needle_character_vec` and `second_needle_character_vec`. Let's also change the comment (l. 573) to `The first/second needle character broadcasted into a 16 bytes vector`.

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -587,9 +589,14 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
 
+#ifdef __SSE4_1__
+        first_pattern = _mm_set1_epi8(first);
+        if (needle + 1 < needle_end)
```
More intuitive: `if (needle_size > 1)`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -569,8 +569,10 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
 #ifdef __SSE4_1__
-    /// vector filled `first` for determining leftmost position of the first symbol
-    __m128i pattern;
+    uint8_t second{};
```
About `#ifdef __SSE4_1__`: As far as I see (--> [Intel Intrinsics Guide](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html)), all used intrinsics in the block l. 593ff and the block l. 671 are SSE2 intrinsic. Could we ifdef for SSE2 instead?

jiebinn replied:
Done.

rschu1ze replied:
The same comment also applies to all other code blocks touched in this PR.

jiebinn replied:
Yes, you're right. I have checked all the intrinsic code part. "_mm_insert_epi8" is from SSE4.1 and all the others are from SSE2. I have pushed a patch to correct it. But I have a question to do that.

The platform from SSE2 to SSE4.1 could not use the SSE4.1 acceleration. The SSE2 part code is useless and waste time for the platform from SSE2 to SSE4.1. All the code block work together for the same goal no matter from the SSE2 or the SSE4.1.
So we might align all the code blocks to SSE4.1?

rschu1ze replied:
Good point, I seemingly missed `_mm_insert_epi8`. In that case, yes, you could align all the code to SSE4.1. It actually doesn't matter since the ClickHouse build itself requires at least SSE 4.2 (file "cmake/cpu_features.cmake")...

jiebinn replied:
If the platform is SSE3, then it will execute all the SSE2 code block. But it still could not use the SIMD acceleration as it need the SSE4.1 code block. And it will fall back to the "Byte-by-Byte comparison". Then the SSE2 code block might waste time for the SSE3 platform?

jiebinn replied:
OK. I have aligned all the code block to SSE4.1.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+                        continue;
+                    }
+
+                    const auto offset = __builtin_ctz(mask);
```
`__builtin_ctz` --> `std::countr_zero()`

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -569,8 +569,10 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+    uint8_t second{};
+    /// vector filled `first` or `second` for determining leftmost position of the first and second symbols
+    __m128i first_pattern;
+    __m128i second_pattern;
```
Unrelated, but a `short` or `uint16_t` would do the job for `cachemask` (l. 578)

jiebinn replied:
Done. uint16_t should be fine for cachemask.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -587,9 +589,14 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+        {
+            second  = *(needle + 1);
+            second_pattern = _mm_set1_epi8(second);
+        }
```
Unrelated, but in l. 602, why not simply `for (int i=0; i<n; ++i)`?

jiebinn replied:
Done. Simplify the original code with a simple loop.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
-                const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, pattern);
+                if (haystack + n <= haystack_end && pageSafe(haystack))
+                {
+                    const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
```
It would be nice to replace `auto` by the actual type (e.g. `_m128i`) because it is usually not immediately clear which return type the intrinsics have.

jiebinn replied:
Done. I have also checked the other values which from intrinsics.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
-        {
 #ifdef __SSE4_1__
-            if (haystack + n <= haystack_end && pageSafe(haystack))
+        /// Here is the quick path when needle_size is 1. Compare the first and second characters if the needle_size >= 2.
```
Let's make the comment simpler and more precise: `/// Fast path for single-character needles. Compare 16 characters of the haystack against the needle character at once.`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
-                const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, pattern);
+                if (haystack + n <= haystack_end && pageSafe(haystack))
+                {
+                    const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
```
`v_haystack` --> `haystack_characters`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+                if (haystack + n <= haystack_end && pageSafe(haystack))
+                {
+                    const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                    const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, first_pattern);
```
`v_against_pattern` --> `comparison_result`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+                {
+                    const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                    const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, first_pattern);
+                    const auto mask = _mm_movemask_epi8(v_against_pattern);
```
`mask` --> `comparison_result_mask`

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+        {
+#ifdef __SSE4_1__
+            /// find first and second characters
+            if ((haystack + 1 + n) <= haystack_end && pageSafe(haystack))
```
Isn't that `pageSafe(haystack + 1)`?

jiebinn replied:
Yes. I think it should be pageSafe(haystack + 1) based on the pagesafe function.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+            /// find first and second characters
+            if ((haystack + 1 + n) <= haystack_end && pageSafe(haystack))
+            {
+                const auto v_haystack_block_first = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
```
`v_haystack_block_first` --> `haystack_characters_from_1st`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+            if ((haystack + 1 + n) <= haystack_end && pageSafe(haystack))
+            {
+                const auto v_haystack_block_first = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                const auto v_haystack_block_second = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack + 1));
```
`v_haystack_block_second` --> `haystack_characters_from_2nd`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+            {
+                const auto v_haystack_block_first = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                const auto v_haystack_block_second = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack + 1));
+                const auto v_against_pattern_first = _mm_cmpeq_epi8(v_haystack_block_first, first_pattern);
```
`v_against_pattern_first` --> `comparison_result_1st`
`v_against_pattern_second` --> `comparison_result_2nd`

jiebinn replied:
Done.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+                const auto v_haystack_block_second = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack + 1));
+                const auto v_against_pattern_first = _mm_cmpeq_epi8(v_haystack_block_first, first_pattern);
+                const auto v_against_pattern_second = _mm_cmpeq_epi8(v_haystack_block_second, second_pattern);
+                const auto mask = _mm_movemask_epi8(_mm_and_si128(v_against_pattern_first, v_against_pattern_second));
```
For debugging, it will be nicer if the result of ANDing both results will be a separate variable (it can be called `comparison_result_combined`

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+                const auto v_against_pattern_first = _mm_cmpeq_epi8(v_haystack_block_first, first_pattern);
+                const auto v_against_pattern_second = _mm_cmpeq_epi8(v_haystack_block_second, second_pattern);
+                const auto mask = _mm_movemask_epi8(_mm_and_si128(v_against_pattern_first, v_against_pattern_second));
+                /// first and second characters not present in 16 octets starting at `haystack`
```
Question: Is above code (l. 706ff.) safe in cases where the haystack contains the first and second needle characters but not consecutively. Example:
- needle = 'ab', 1st needle character: 'a', 2nd needle character: 'b'
- haystack = 'xx a yy b zz'.

jiebinn replied:
In this example,

first_needle_character_vec = aaaaaaa
second_needle_character_vec = bbbbbbb
haystack_characters_from_1st = xxayybz
haystack_characters_from_2nd = xayybzz
comparison_result_1st = 0010000
comparison_result_2nd = 0000100
comparison_result_combined = 0000000

Then there will no matched 'ab' in haystack.

rschu1ze replied:
You are right, thanks.

rschu1ze commented on src/Common/StringSearcher.h:
```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
+        while (haystack < haystack_end)
+        {
+#ifdef __SSE4_1__
+            /// find first and second characters
```
Could we make this comment a bit more useful?

jiebinn replied:
Done. I have added several detailed comments.

rschu1ze commented on src/Common/Volnitsky.h:
```diff
@@ -429,6 +433,10 @@ class VolnitskyBase
 
         const auto * haystack_end = haystack + haystack_size;
 
+#ifdef __SSE4_1__
```
Are you sure you want to invoke the fallback searcher (instead of Volnitsky) for *all* systems with SSE 4.1. That are pretty much all machines nowadays which means Volnitsky is no longer called at all.

jiebinn replied:
The Volnitsky perform good when the needle_size is very large, but not good using small needle_size. 

I have pushed another commit 'Optimize the SIMD StringSearcher if needle_size is large'. Currently, the optimized SIMD has win the Volnitsky almost all the time. The optimized SIMD StringSearcher has led at most 41.7% than Volnitsky algorithm when the needle_size is less than 21, and fallen behind only about 1% even when the needle_size is bigger than 50, which is not considered as a common case.

If the platform support X86 SSE, we could use the optimized SIMD StringSearcher.
If the platform does not support X86 SSE, then we should Use Volnitsky or the common Byte-by-Byte search (fallback). 

Test platform: ICX server
Test query: SELECT COUNT(*) FROM hits WHERE URL LIKE '%{Needle}%';

Needle_size	opt/baseline
5		141.7%
6		129.4%
8		118.5%
9		112.3%
10		107.4%
14		103.4%
20		100.2%
21		100.7%
51		99.0%

rschu1ze replied:
Very interesting numbers, thank you. In the PR, you wrote that Volnitsky wins if the needle is very long (e.g. > 150 characters). Does it make sense to make the fallback choice dependent on the needle length?

jiebinn commented on src/Common/StringSearcher.h:
```diff
@@ -4,6 +4,7 @@
 #include <Common/Exception.h>
 #include <Common/StringUtils/StringUtils.h>
 #include <Common/UTF8Helpers.h>
+#include <Common/TargetSpecific.h>
```
There is no need to use TargetSpecific.h for SSE code. I will leave the AVX512 optimization for another PR.

jiebinn commented:
Hi @rschu1ze , thanks for your advice. I have git commit a patch based on your suggestion, which make the code clean and add some comments. 
Besides, I have also pushed another patch which improve the performance when needle_size is large. Currently, the optimized SIMD StringSearcher could lead than Volnitsky almost all the time.

rschu1ze commented:
After reading the PR in detail, it looks really good. Thanks! I just left a minor comment, the PR should be ready for merge then. (We should also check the performance test results from CI.) And before you push more optimizations (I read "AVX" somewhere) - I have a small cleanup PR which should go in afterwards.

jiebinn commented:
> After reading the PR in detail, it looks really good. Thanks! I just left a minor comment, the PR should be ready for merge then. (We should also check the performance test results from CI.) And before you push more optimizations (I read "AVX" somewhere) - I have a small cleanup PR which should go in afterwards.

Thanks. I am working in the AVX512 optimizations for this part as AVX512 has new intrinsics to boost and also make some quick path. But I still need more time to collect performance data and polish the draft patch based on the hotspot function. I will let you know if I finish it. It should be another PR.

jiebinn committed:
Optimize the SIMD StringSearcher if needle_size is large

This patch offers an additional optimization when the needle_size is
large. If the needle_size is larger than the haystack_size, there is
no need to search any more.

The optimized SIMD StringSearcher has led at most 41.7% than Volnitsky
algorithm when the needle_size is less than 21, and fallen behind only
about 1% even when the needle_size is bigger than 50, which is not
considered as a common case.

Test platform: ICX server
Test query: SELECT COUNT(*) FROM hits WHERE URL LIKE '%{Needle}%';

Needle_size	opt/baseline
5		141.7%
6		129.4%
8		118.5%
9		112.3%
10		107.4%
14		103.4%
20		100.2%
21		100.7%
51		99.0%

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Make the optimized SIMD StringSearcher clean

This patch has revised the name of value and added comments to make
the SIMD StringSearcher clean and easy to understand based on pull
request 46289.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

jiebinn committed:
Align all the SSE4.1 requirement and use needle_size

Align all the SSE4.1 requirement from StringSearcher. Use needle_size
in while loop to make the code clean.

Signed-off-by: Jiebin Sun <jiebin.sun@intel.com>

rschu1ze commented:
- The crash in [ClickHouse Stress Test (debug)](https://s3.amazonaws.com/clickhouse-test-reports/46289/d6710d9b34ed223cd8306c738b9b05186b6e9172/stress_test__debug_.html) looks like it is fixed in upstream ClickHouse by https://github.com/ClickHouse/ClickHouse/pull/46600 / https://github.com/ClickHouse/llvm-project/pull/8 
- whereas the failure in [ClickHouse Stress Test (msan)](https://s3.amazonaws.com/clickhouse-test-reports/46289/d6710d9b34ed223cd8306c738b9b05186b6e9172/stress_test__msan_.html) is unrelated and
- I have seen most of the crashes in [ClickHouse Stress Test (ubsan)](https://s3.amazonaws.com/clickhouse-test-reports/46289/d6710d9b34ed223cd8306c738b9b05186b6e9172/stress_test__ubsan_.html) also in another PR.

But to be sure, I'll restart these tests because memory corruption due to SIMD can be nasty to debug.

Btw., one search-related query [in the performance tests](https://s3.amazonaws.com/clickhouse-test-reports/46289/d6710d9b34ed223cd8306c738b9b05186b6e9172/performance_comparison_[4/4]/report.html) became faster (good!).

rschu1ze committed:
Merge branch 'master' into StringSearcher

jiebinn commented:
@rschu1ze Nice to see a boost of more than one third!

rschu1ze commented:
ClickHouse Stress Test (ubsan):
ClickHouse Stress Test (debug)
- fatal_messages.txt: Crashes also occur in branch master, unrelated.

ClickHouse Stress Test (msan):
- stderror.log:
```
==1569==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4b98957c in DB::TCPHandler::sendLogs() build_docker/../src/Server/TCPHandler.cpp:1943:9
    #1 0x4b9ada52 in DB::TCPHandler::runImpl()::$_2::operator()() const build_docker/../src/Server/TCPHandler.cpp:304:21
```
Looks like a false positive or perhaps variable `logs_queue` must be accessed under a lock. In any case, unrelated.
- fatal_messages.txt: Crashes also occur in branch master, unrelated.

One performance test didn't run (also due to unrelated reasons). [This performance test](https://s3.amazonaws.com/clickhouse-test-reports/46289/6aa4e93284eba06049b994da2cdeb8f446201d7c/performance_comparison_[4/4]/report.html) confirms the speedups (in this case for `match()` predicates)

rschu1ze commented:
@jiebinn If you plan to push more optimizations along these lines, feel free to assign me as reviewer directly, thanks.

jiebinn commented:
Sure. Thanks for the review.
